server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -114,11 +140,36 @@ void update(ShardRouting oldShard, ShardRouting newShard) {
         }
         ShardRouting previousValue = shards.put(newShard.shardId(), newShard);
         assert previousValue == oldShard : "expected shard " + previousValue + " but was " + oldShard;
+
+        if (oldShard.initializing()) {
+            boolean exist = initializingShards.remove(oldShard);
+            assert exist : "expected shard " + oldShard + "exists in initializingShards, but not";
+        } else if (oldShard.relocating()) {
+            boolean exist = relocatingShards.remove(oldShard);
+            assert exist : "expected shard " + oldShard + "exists in relocatingShards, but not";
+        }
+
+        if (newShard.initializing()) {
+            initializingShards.add(newShard);
+        } else if (newShard.relocating()) {
+            relocatingShards.add(newShard);
+        }
+
+        assert invariant();
     }
 
     void remove(ShardRouting shard) {
         ShardRouting previousValue = shards.remove(shard.shardId());
         assert previousValue == shard : "expected shard " + previousValue + " but was " + shard;
+        if (shard.initializing()) {
+            boolean exist = initializingShards.remove(shard);
+            assert exist : "expected shard " + shard + "exists in initializingShards, but not ";
```

Comment:
Wording/whitespace suggestion:

```suggestion
            assert exist : "expected shard " + shard + " to exist in initializingShards";
```